### PR TITLE
 Move Concat balancing from DFG to FuncOpt

### PR DESCRIPTION
**Pull request by gezalore**
This is better done in FuncOpt as well, as it applies in more cases, outside combinational logic which is non represent able in DFG.

This is an effective revert of #5598, and adding back the same logic in FuncOpt.

### CODE REVIEW

**Comment by wsnyder** [comment]
Seems fundamentally good.  Will give it another quick review once  #5599 is in master and you merge master here, so I can see just the deltas (unless you plan to merge differently?).

**Comment by gezalore** [comment]
Will merge after #5599, so it's Ok to wait.

**Comment by gezalore** [comment]
Now rebased.

**Review by wsnyder — APPROVED** [review]

**Pull request merged**
